compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -256,22 +256,22 @@ pub fn each_linked_rlib(
         Some(f) => f,
         None => return Err("could not find formats for rlibs".to_string()),
     };
-    for &(cnum, ref path) in crates {
+    for &cnum in crates {
         match fmts.get(cnum.as_usize() - 1) {
             Some(&Linkage::NotLinked | &Linkage::IncludedFromDylib) => continue,
             Some(_) => {}
             None => return Err("could not find formats for rlibs".to_string()),
         }
         let name = &info.crate_name[&cnum];
-        let path = match *path {
-            LibSource::Some(ref p) => p,
-            LibSource::MetadataOnly => {
-                return Err(format!(
-                    "could not find rlib for: `{}`, found rmeta (metadata) file",
-                    name
-                ));
-            }
-            LibSource::None => return Err(format!("could not find rlib for: `{}`", name)),
+        let path = if let Some((path, _)) = &info.used_crate_source[&cnum].rlib {
+            path
+        } else if info.used_crate_source[&cnum].rmeta.is_some() {
```

Comment:
The `info.used_crate_source[&cnum]` search can be done once instead of twice here.